✨ 인증 서비스 토큰 검증 내부 통신 API 추가

tellang opened the pull request:
## 🔗 관련 이슈

Closes #68

## 📌 PR 요약

- JWT 토큰 검증 기능 추가 및 관련 예외 처리 기능 구현

## 🔍 주요 변경 사항

- `CustomException`에 `InvalidTokenException` 클래스 추가
  - `ErrorCode`에 `INVALID_TOKEN` 코드 및 메시지 추가
- `AuthService`에 JWT 토큰 검증 로직 추가
  - `AuthInternalController`에 JWT 검증 엔드포인트 추가
  - `build.gradle`에 Spring WebFlux 및 Reactor 테스트 의존성 추가
- `AuthServiceTest`에 JWT 토큰 검증 테스트 케이스 추가

## 🧪 테스트 방법
```bash
./gradlew test
```

## ✅ PR 체크리스트

- [x] 코드 컨벤션을 준수했나요?
- [x] 필요한 테스트를 추가하고 모든 테스트가 통과하나요?
- [ ] 관련 문서를 업데이트했나요? (필요한 경우)

## 👀 리뷰어 체크 포인트


## 💬 기타 코멘트

- 공통 모듈로 통합 예정인 코드가 있습니다
  - 예외 처리 핸들링의 경우 공통 모듈 통합시 적용예정입니다

tellang committed:
:sparkles: `CustomException`에 `InvalidTokenException` 클래스 추가 및 `ErrorCode`에 관련 코드 추가 (#68)

- 공통 모듈에 통합 예정
- JWT 토큰이 유효하지 않을 때 발생하는 `InvalidTokenException` 클래스 추가
- `ErrorCode`에 `INVALID_TOKEN` 코드 및 메시지 추가

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:sparkles: JWT 토큰 검증 기능 추가 (#68)

- `AuthService`에 JWT 토큰 검증 메서드 추가
- `AuthInternalController` 생성 및 JWT 토큰 검증 엔드포인트 추가
- `build.gradle`에 Spring WebFlux 의존성 및 Reactor 테스트 지원 의존성 추가

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:white_check_mark: `AuthServiceTest`에 JWT 토큰 검증 테스트 추가 (#68)

- 유효한 JWT 토큰 검증 및 클레임 정보 반환 테스트 추가
- 유효하지 않은 JWT 토큰 검증 시 예외 발생 테스트 추가

Signed-off-by: tellang <tellang@daum.net>

SerenityZenDev commented on auth-service/build.gradle:
```diff
@@ -7,6 +7,9 @@ plugins {
 
 dependencies {
+    // Spring WebFlux 의존성 추가
+    implementation 'org.springframework.boot:spring-boot-starter-webflux'
```
인증 서비스를 비동기로 변경하는 것으로 보이는데 유저 서비스가 동기인 상태로 진행하면 문제되는 점은 없나요?

tellang replied:
JWT 유효성 검사 로직이 GW에서 비동기적으로 Auth 서비스의 내부 API를 호출하는 구조입니다.  

처음에는 `GW에서 비동기 호출을 하니 Auth 서비스도 비동기로 구현해야 한다`는 다소 나이브한 생각으로 비동기 방식을 채택했습니다.  

그러나 이 부분에서는 좀 더 신중한 접근이 필요할 것 같습니다.

tellang replied:
조금 더 고민해 보면 GW와 Auth 서비스의 관계를 비동기로 처리하는 것이 이번 프로젝트의 목적에 더 부합할 것 같습니다. 

GW는 많은 트래픽을 처리해야 하기 때문에 **비동기 방식**으로 Auth 서비스에 요청을 빠르게 던진 후  

Auth 이후의 작업을 처리하는 것이 효율적이라고 생각이 듭니다



반면, Auth 서비스 이후의 User 서비스는 동기적으로 처리되기 때문에 

Auth와 User 간의 처리과정의 고려할 사항들을 생각해보면

**Auth와 User 서비스** 사이에서는 **동기 방식**으로 진행하는 방향이 현재 상황에선 좋은 선택인 것 같습니다!

tellang committed:
:recycle: `JwtUtil` 클래스 리팩토링 및 `validateToken` 메서드 개선 (#68)

- `JwtUtil` 클래스에 JWT 토큰 유효성 검증 및 클레임 추출을 결합한 `validateAndExtractClaims` 메서드 추가
  - 기존 중복된 `extractClaims` 호출방식을 방식을 해소

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:zap:  AuthService 및 AuthServiceTest 비동기에서 동기 방식으로 변경 (#68)

- AuthService의 Mono 기반 비동기 메서드를 동기 방식으로 변경
  - `build.gradle`에 Spring WebFlux 의존성 및 Reactor 테스트 지원 의존성 제거
  - validateToken 메서드를 동기 처리 방식으로 수정

- AuthServiceTest에서 비동기 테스트 로직 제거 및 동기 방식으로 수정
  - StepVerifier 제거
  - Mono 반환 타입을 Map으로 변경
  - JWT 토큰 검증 및 클레임 반환 테스트 수정
  - 비밀번호 불일치 예외 발생 테스트 유지

Signed-off-by: tellang <tellang@daum.net>

SerenityZenDev reviewed (approved):
제가 webflux 의존성에 comment를 남겨서 아직 보이는건지 모르겠는데 동기방식으로 변경된것으로 확인됩니다. 수고하셧습니다
